### PR TITLE
Advance the clock through split-explicit barotropic substeps

### PR DESCRIPTION
**Pull request by enri66**
## Summary

Follow-up to [Simone's comment on #5970](https://github.com/CliMA/Oceananigans.jl/pull/5970#issuecomment-5647126382) (👍'd there as a separate PR).

The clock used to fill `U`, `V` and `η` halos during split-explicit barotropic subcycling was built once before the substep loop, at the macro time step's starting time, and reused unchanged for every substep. A time-dependent barotropic boundary condition (e.g. a tidal Flather condition) therefore saw the same stale time on every one of the ~30-50 substep evaluations instead of the time actually advancing through them.

Now a fresh clock is built each substep from the fixed barotropic `Δτ` and spliced into the pre-converted halo-fill arguments right before the last of them (`barotropic_model_fields`), the same lightweight way `averaging_weight`/`transport_weight` are already threaded per substep — so the argument-conversion hoisting this loop relies on for performance (documented in the existing comment above the loop) is untouched.

## Test plan

- [x] New test in `test/test_split_explicit_free_surface_boundaries.jl`: a time-dependent `GravityWaveRadiationBoundaryCondition` records every clock time it's evaluated at during one time step; asserts more than one distinct time is seen, evenly spaced by the barotropic `Δτ`. Verified this test fails against the pre-fix code (`length(substep_times) > 1` fails with exactly 1).
- [x] Full `test_split_explicit_free_surface_boundaries.jl` suite passes (52/52 existing + 2/2 new).
- [x] `hydrostatic_free_surface` test group passes (339 passed, 7 known-broken unchanged, 0 failed).
- [x] `hydrostatic_regression` test group passes (48/48, residuals ~1e-7, unchanged from baseline).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by simone-silvestri** [comment]
Looks good, I enabled the automatic benchmarking, since the substepping is a delicate performance hotspot (it is negligible now but some inference error could blow up the performance given that we have many substeps).
Once the benchmarks pass this looks good to merge.

@enri66 you need to commit to enable the benchmarks run (`git commit --allow-empty -m "restart CI"` then `git push`)

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl [hunk 1]
```diff
@@ -128,11 +128,12 @@ function iterate_split_explicit!(free_surface::FillHaloSplitExplicit, grid, GU
...
     barotropic_model_fields = (; U, V, η)
 
-    # a substep clock with a smaller Δτ is needed for inter-step boundary conditions to be valid
-    substep_clock = (; time = clock.time, iteration = clock.iteration, stage = 0, last_stage_Δt = Δτᴮ)
-    @apply_regionally U_halo_args = build_halo_fill_args(U, grid, substep_clock, barotropic_model_fields)
-    @apply_regionally V_halo_args = build_halo_fill_args(V, grid, substep_clock, barotropic_model_fields)
-    @apply_regionally η_halo_args = build_halo_fill_args(η, grid, substep_clock, barotropic_model_fields)
+    # Boundary conditions on U, V and η are time-dependent, so each substep needs its own clock;
+    # built fresh every iteration below and spliced in just before `barotropic_model_fields`, which
+    # is always the last element regardless of grid type (see `build_halo_fill_args`).
```
Comment (simone-silvestri) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl]:
```suggestion
    # Builds also a separate "sub-stepping" clock to account for time dependent forcing and boundary conditions
```

**Comment by simone-silvestri** [comment]
hmmm, looks like the benchmarks pipeline does not run on forks. Oh well, we will see the effect on main as the PR gets merged.

**Comment by enri66** [comment]
Applied, and pushed the restart-CI commit for the benchmarks.

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl [hunk 2]
```diff
@@ -149,11 +148,13 @@ function iterate_split_explicit!(free_surface::FillHaloSplitExplicit, grid, GU
             @inbounds averaging_weight = weights[substep]
             @inbounds transport_weight = transport_weights[substep]
 
-            maybe_distributed_fill_halo_regions!(arch, converted_η_halo_args...; only_local_halos)
+            substep_clock = (; time = clock.time + (substep - 1) * Δτᴮ, iteration = clock.iteration, stage = 0, last_stage_Δt = Δτᴮ)
```
Comment (simone-silvestri) [inline src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl:151]: actually I wonder, instead of building this object in the substepping, can we build it ouside as we do now and tick the "time" value in the substeps? I am not sure if this would lead to less allocation

**Comment by enri66** [comment]
On the `codecov/patch` failure: the 4 "missing" lines are the three `@apply_regionally U/V/η_halo_args = build_halo_fill_args(...)` setup calls just before the substep loop (lines 132-134). The lines that actually matter for this fix — the new `substep_clock` construction and the three `maybe_distributed_fill_halo_regions!` calls inside the loop (151, 153, 156, 157) — all show as hit, so the regression test does exercise the fix.

The `@apply_regionally` lines were already showing 0 coverage on `main` before this PR touched them, with identical content — this looks like a pre-existing quirk with Julia's line-coverage instrumentation and the `@apply_regionally` multi-region dispatch macro, not something introduced here. Not something to fix in this PR.

**Comment by simone-silvestri** [comment]
@enri66 if you resolve the conflicts I will merge when tests pass

File: src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl (hunk 2 above)
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/step_split_explicit_free_surface.jl:152]: @enri66 just if maybe you can address this comment

**Comment by enri66** [comment]
@simone-silvestri about the comment from last week:  `substep_clock` is `isbitstype`, so each reconstruction is a stack value, not a heap allocation. I benchmarked `time_step!` on this PR's own test model (`substeps=100`, with a boundary condition that reads `clock.time` on every substep) against an identical model with a time-independent boundary condition — allocations were identical (538512 bytes either way), so rebuilding it 100×/step costs nothing measurable.

This is also exactly the existing pattern elsewhere: `Clock` is a `mutable struct`, and its own `Adapt.adapt_structure`/`convert_time` explicitly rebuild it into a fresh *immutable* NamedTuple at every kernel-launch site, since kernel arguments need to be immutable/isbits — a mutated, "ticked" clock wouldn't be a valid substitute. So I'd leave this as is.

**Comment by enri66** [comment]
@simone-silvestri `cpu-all-tests` failed here too, but on a different error than #5970's — a filesystem race during precompilation:

```
ERROR: LoadError: SystemError: opening file "/var/lib/buildkite-agent/.julia/compiled/v1.12/Oceananigans/hU93i_wbkQg.ji": No such file or directory
```

That's a parallel test-worker cache collision, not a real test failure — nothing in this PR's diff (the substep clock fix) is anywhere near it. Combined with #5970's unrelated `ExplicitImports` failure, this looks like general CI flakiness from the `ParallelTestRunner` migration rather than anything code-side. `oceananigans-dot-jl` (AMDGPU) passed cleanly here, for what it's worth.

Same ask as #5970 — could you kick off a rebuild when you get a chance?
